### ISSUE TITLE
Module level fixture log capture and association

### ISSUE THREAD

**Issue by leezu** [pytest-dev/pytest#7819]
There are two issues with pytest implementation of module level fixtures and logging statements therein:

1) Any captured log is associated with the first test of the module and thus generally never printed as long as the first test of the module passes
2) It is impossible to disable capturing of logging messages in module level fixtures via `-s`

Consider conftest

``` python
import logging
import pytest

@pytest.fixture(scope="module", autouse=True)
def module_fixture():
    logging.warning('LOGGING')
    print('PRINT')
```

and test file

``` python
import pytest

def test_pass():
    pass

def test_fail():
    pytest.fail()
```

We can run the pytest with default options, the module level fixtures are captured and not printed.

```
% python3.8 -m pytest .                                                                                                                                                         ~/test
================================================================================= test session starts =================================================================================
platform linux -- Python 3.8.0, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/ANT.AMAZON.COM/lausen/test
plugins: env-0.6.2, flaky-3.6.1
collected 2 items

test_logging.py .F                                                                                                                                                              [100%]

====================================================================================== FAILURES =======================================================================================
______________________________________________________________________________________ test_fail ______________________________________________________________________________________

    def test_fail():
>       pytest.fail()
E       Failed: <Failed instance>

test_logging.py:7: Failed
=============================================================================== short test summary info ===============================================================================
FAILED test_logging.py::test_fail - Failed: <Failed instance>
============================================================================= 1 failed, 1 passed in 0.05s =============================================================================
```

We can pass `-s` and print statement to stdout is no longer captured. The `logging.warning` statement is still captured and never shown:

```
% python3.8 -m pytest -s .                                                                                                                                                      ~/test
================================================================================= test session starts =================================================================================
platform linux -- Python 3.8.0, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/ANT.AMAZON.COM/lausen/test
plugins: env-0.6.2, flaky-3.6.1
collected 2 items

test_logging.py PRINT
.F

====================================================================================== FAILURES =======================================================================================
______________________________________________________________________________________ test_fail ______________________________________________________________________________________

    def test_fail():
>       pytest.fail()
E       Failed: <Failed instance>

test_logging.py:7: Failed
=============================================================================== short test summary info ===============================================================================
FAILED test_logging.py::test_fail - Failed: <Failed instance>
============================================================================= 1 failed, 1 passed in 0.05s =============================================================================
```

If we swap the order of the tests in the testfile, the captured content is shown:

```
% python3.8 -m pytest .                                                                                                                                                         ~/test
================================================================================= test session starts =================================================================================
platform linux -- Python 3.8.0, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/ANT.AMAZON.COM/lausen/test
plugins: env-0.6.2, flaky-3.6.1
collected 2 items

test_logging.py F.                                                                                                                                                              [100%]

====================================================================================== FAILURES =======================================================================================
______________________________________________________________________________________ test_fail ______________________________________________________________________________________

    def test_fail():
>       pytest.fail()
E       Failed: <Failed instance>

test_logging.py:4: Failed
-------------------------------------------------------------------------------- Captured stdout setup --------------------------------------------------------------------------------
PRINT
--------------------------------------------------------------------------------- Captured log setup ----------------------------------------------------------------------------------
WARNING  root:conftest.py:6 LOGGING
=============================================================================== short test summary info ===============================================================================
FAILED test_logging.py::test_fail - Failed: <Failed instance>
============================================================================= 1 failed, 1 passed in 0.06s =============================================================================
```

**Comment by bluetech** [comment]
Hi @leezu,

> Any captured log is associated with the first test of the module and thus generally never printed as long as the first test of the module passes

That's how pytest works. Fixtures are executed "lazily", i.e. when something requires them, and all fixtures are executed in the context of some test. There are some trade-offs, but generally it is unlikely to change.

> It is impossible to disable capturing of logging messages in module level fixtures via -s

Output capturing (`-s`) and log capturing are separate matters. They are related in that a log message might print to stdout/stderr but log capturing itself doesn't work at that level, it works by attaching a handler to the logging mechanism itself.

In light of that, can you rephrase the question? It (and the example) are not entirely clear otherwise.

**Comment by leezu** [comment]
Thanks @bluetech.

> Fixtures are executed "lazily", i.e. when something requires them, and all fixtures are executed in the context of some test. There are some trade-offs, but generally it is unlikely to change.

There are also `autouse=True` fixtures which are always executed at the beginning of the scope (eg module).
It makes sense to associated `scope=function` fixtures with the test itself, but it doesn't appear correct to handle fixtures with different scope in the same manner. We should associate a fixture with all functions that are part of it's scope.

> Output capturing (-s) and log capturing are separate matters. They are related in that a log message might print to stdout/stderr but log capturing itself doesn't work at that level, it works by attaching a handler to the logging mechanism itself.
> In light of that, can you rephrase the question? It (and the example) are not entirely clear otherwise.

Thank you for clarifying. My question is, why does `pytest --show-capture=no` not show the logging message (`WARNING  root:conftest.py:6 LOGGING`) when it is logged, but capture it nevertheless. Ie. how to disable log capturing?

I'm interested in disabling log capturing so that I can see log messages printed by a module scope fixture, which are currently captured and never shown to the user because the module fixture is associated with a passing test.

**Comment by bluetech** [comment]
Your feature request sort-of makes sense. I think it would be interesting to think how it should work though. For example if multiple tests in the module fail, should the module-level log messages be duplicated in each test's report, or somehow be displayed only once? etc. In any case it would be quite a challenge to implement.

It can be approximated by using a function-scope fixture F which depends on the module-scope fixture M and have F print M's log messages to stderr or such. Then it will be displayed in each failing function's report by default.

> There are also autouse=True fixtures which are always executed at the beginning of the scope (eg module).

`autouse` just means each test requests the fixture implicitly.

> We should associate a fixture with all functions that are part of it's scope.

Perhaps, this is not something I have thought through myself, but as I said it is not very likely to change, unless someone is willing to put a lot of design and implementation work on that (if it is indeed better).

> why does pytest --show-capture=no not show the logging message (WARNING root:conftest.py:6 LOGGING) when it is logged, but capture it nevertheless.

`--show-capture` only controls what is displayed but not whether log capturing itself happens.

> Ie. how to disable log capturing?

Nuclear option is to disable the logging plugin entirely using `-p no:logging`. There is no specific flag for this.

A more fine-grained option is to use the `caplog` fixture inside a test or a fixture.

**Comment by leezu** [comment]
> For example if multiple tests in the module fail, should the module-level log messages be duplicated in each test's report, or somehow be displayed only once? etc.

I think it's ok to display the module-level log for each failed test. 

Thank you for elaborating on the logging plugin configuration and caplog.

**Comment by vitaly-krugl** [comment]
I just ran into this myself with class-level fixtures and I am stumped. Some of my tests failed, and I suspect it's related to something not going right in the fixture, but the log from the fixture is absent from the logs of the failed tests.

When debugging failed tests, what happens in the fixture is incredibly important (otherwise why would you have the fixtures), so any logging output of the fixtures (class, module, method, etc.) should be present alongside the other log output of each failed test to help with debugging.
